src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -0,0 +1,563 @@
+                        if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+                        return actual;
+                    }
+
+                    if (result.IsCompleted)
+                    {
+                        if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+                        return 0;
+                    }
+                }
+                finally
+                {
+                    _requestPipe.Reader.AdvanceTo(consumed);
+                }
+            }
+        }
+
+        internal override void ShutdownRead()
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            _requestPipe.Reader.Complete();
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+        }
+
+        internal override void ShutdownWrite()
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            // TODO do anything to stop writes?
+            _api.StreamShutdownDelegate(_ptr, (uint)QUIC_STREAM_SHUTDOWN_FLAG.GRACEFUL, 0);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+
+        }
+
+        internal override int Read(Span<byte> buffer)
+        {
+            throw new NotImplementedException();
+        }
+
+        internal override void Write(ReadOnlySpan<byte> buffer)
+        {
+            throw new NotImplementedException();
+        }
+
+        internal override void Flush()
+        {
+            throw new NotImplementedException();
+        }
+
+        internal override async Task FlushAsync(CancellationToken cancellationToken)
+        {
+            await SendAsync(new ReadOnlySequence<byte>(Memory<byte>.Empty), QUIC_SEND_FLAG.NONE);
+        }
+
+        public override async ValueTask DisposeAsync()
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            if (_ptr != IntPtr.Zero)
+            {
+                await ShutdownAsync(QUIC_STREAM_SHUTDOWN_FLAG.GRACEFUL, 0).ConfigureAwait(false);
+                _api.StreamCloseDelegate?.Invoke(_ptr);
+            }
+
+            _handle.Free();
+            _api = null;
+
+            _disposed = true;
+        }
+
+        public override void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~MsQuicStream()
+        {
+            Dispose(false);
+        }
+
+        // Synchronous shutdown current does a graceful shutdown, which must go async
+        // Close can be done synchronously, but there is not guarantee that all data will be sent to the client
+        // We probably need to reconsider how to handle dispose/shutdown cases.
+        private void Dispose(bool disposing)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            if (_ptr != IntPtr.Zero)
+            {
+                ShutdownAsync(QUIC_STREAM_SHUTDOWN_FLAG.GRACEFUL, 0).GetAwaiter().GetResult();
```

Comment:
😢

Reply:
But we have `DisposeAsync`.

Reply:
I think I'm going to make this abort instead of close gracefully